# Sprint 3 M3 T02

## Ejercicio 1

### 1. Descarga el conjunto de datos 'Airlines Delay: Airline on-time statistics and delay causes' y cárgalo en un marco de datos de Pandas.

Se importa la librería de Pandas:

In [1]:
import pandas as pd

Se descarga el conjunto de datos 'Airlines Delay: Airline on-time statistics and delay causes':

In [2]:
pd.read_csv('DelayedFlights.csv')

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0         2211.0        2225            WN  ...     4.0     8.0          0   
1         1002.0        1000            WN  ...     5.0    10.0          0   
2          804.0         750            WN  ...     3.0    17.0          0   
3         1959.0        1925            WN  ...     3.0    10.0          0   
4         2121.0        2110            WN  ...     4.0    10.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   
1936756   1446.0        1437            DL  ...    13.0    13.0          0   
1936757   1413.0        1418            DL  ...     8.0    11.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                       N         0           NaN           NaN      NaN   
1                       N         0           NaN           NaN      NaN   
2                       N         0           NaN           NaN      NaN   
3                       N         0           2.0           0.0      0.0   
4                       N         0           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   
1936756                 N         0           NaN           NaN      NaN   
1936757                 N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
0                 NaN                NaN  
1                 NaN                NaN  
2                 NaN                NaN  
3                 0.0               32.0  
4                 NaN                NaN  
...               ...                ...  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  
1936756           NaN                NaN  
1936757           NaN                NaN  

[1936758 rows x 30 columns]

Se crea un data frame con el archivo descargado y se usa la función `set_option` para visualizar todas las columnas de una forma fluida y sin interrupciones:

In [3]:
df_flights = pd.read_csv('DelayedFlights.csv')
pd.set_option('display.max_columns', None)

### 2. Explora los datos que contiene el data set y explica brevemente qué variables contiene.

Para saber que datos contiene el archivo y poder verlos más claros, se usan las funciones `head()`, `tail()`, `shape`, `columns` e `info()`.

In [4]:
df_flights.head()

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0           0  2008      1           3          4   2003.0        1955   
1           1  2008      1           3          4    754.0         735   
2           2  2008      1           3          4    628.0         620   
3           4  2008      1           3          4   1829.0        1755   
4           5  2008      1           3          4   1940.0        1915   

   ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
0   2211.0        2225            WN        335  N712SW              128.0   
1   1002.0        1000            WN       3231  N772SW              128.0   
2    804.0         750            WN        448  N428WN               96.0   
3   1959.0        1925            WN       3920  N464WN               90.0   
4   2121.0        2110            WN        378  N726SW              101.0   

   CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  \
0           150.0    116.0     -14.0       8.0    IAD  TPA       810     4.0   
1           145.0    113.0       2.0      19.0    IAD  TPA       810     5.0   
2            90.0     76.0      14.0       8.0    IND  BWI       515     3.0   
3            90.0     77.0      34.0      34.0    IND  BWI       515     3.0   
4           115.0     87.0      11.0      25.0    IND  JAX       688     4.0   

   TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  \
0      8.0          0                N         0           NaN           NaN   
1     10.0          0                N         0           NaN           NaN   
2     17.0          0                N         0           NaN           NaN   
3     10.0          0                N         0           2.0           0.0   
4     10.0          0                N         0           NaN           NaN   

   NASDelay  SecurityDelay  LateAircraftDelay  
0       NaN            NaN                NaN  
1       NaN            NaN                NaN  
2       NaN            NaN                NaN  
3       0.0            0.0               32.0  
4       NaN            NaN                NaN

In [5]:
df_flights.tail()

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  \
1936753   1617.0        1552            DL       1621  N938DL   
1936754    904.0         749            DL       1631  N3743H   
1936755   1149.0        1010            DL       1631  N909DA   
1936756   1446.0        1437            DL       1639  N646DL   
1936757   1413.0        1418            DL       1641  N908DL   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
1936753              147.0           152.0    120.0      25.0      30.0   
1936754              127.0           109.0     78.0      75.0      57.0   
1936755              162.0           143.0    122.0      99.0      80.0   
1936756              115.0           117.0     89.0       9.0      11.0   
1936757              123.0           135.0    104.0      -5.0       7.0   

        Origin Dest  Distance  TaxiIn  TaxiOut  Cancelled CancellationCode  \
1936753    MSP  ATL       906     9.0     18.0          0                N   
1936754    RIC  ATL       481    15.0     34.0          0                N   
1936755    ATL  IAH       689     8.0     32.0          0                N   
1936756    IAD  ATL       533    13.0     13.0          0                N   
1936757    SAT  ATL       874     8.0     11.0          0                N   

         Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
1936753         0           3.0           0.0       0.0            0.0   
1936754         0           0.0          57.0      18.0            0.0   
1936755         0           1.0           0.0      19.0            0.0   
1936756         0           NaN           NaN       NaN            NaN   
1936757         0           NaN           NaN       NaN            NaN   

         LateAircraftDelay  
1936753               22.0  
1936754                0.0  
1936755               79.0  
1936756                NaN  
1936757                NaN

In [6]:
df_flights.shape

(1936758, 30)

In [7]:
df_flights.columns

Index(['Unnamed: 0', 'Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime',
       'CRSDepTime', 'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum',
       'TailNum', 'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
       'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')

In [8]:
df_flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

Ahora sabemos que el dataset contiene 30 columnas y 1936758 filas con datatypes entre objetos, floats e intergers.  

La información es sobre los retrasos de los vuelos (_ArrDelay, DepDelay, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDela_) en Estados Unidos (_Origin, Dest_) del año 2008 entre el 1 de enero y el 13 de diciembre (_Year, Month, DayofMonth, DayOfWeek_).

### 3. Quédate solo con las columnas que consideres relevantes y justifica tu elección.

Se crea un subset quitando las siguientes columnas:

* Unnamed
* Year
* Month
* Day of Month
* Day of Week

Estas columnas con fecha se suprimieron pues son irrelevantes para los proximos ejercicios y la columna Unnamed es una columna extra que no aporta nada.

* Cancellation Code
* Diverted

Las columnas de _CancellationCode_ y _Diverted_ también se eliminaron pues es información que aunque importante en otros contextos, no es relevante para el análisis que haremos más adelante.

In [9]:
df_flights_new = df_flights[[#'Unnamed: 0', 'Year', 'Month', 'DayofMonth', 'DayOfWeek', 
    'DepTime', 'CRSDepTime', 'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum',
       'TailNum', 'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut', 'Cancelled',
    #'CancellationCode', 'Diverted', 
    'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']]
df_flights_new

DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum  \
0         2003.0        1955   2211.0        2225            WN        335   
1          754.0         735   1002.0        1000            WN       3231   
2          628.0         620    804.0         750            WN        448   
3         1829.0        1755   1959.0        1925            WN       3920   
4         1940.0        1915   2121.0        2110            WN        378   
...          ...         ...      ...         ...           ...        ...   
1936753   1250.0        1220   1617.0        1552            DL       1621   
1936754    657.0         600    904.0         749            DL       1631   
1936755   1007.0         847   1149.0        1010            DL       1631   
1936756   1251.0        1240   1446.0        1437            DL       1639   
1936757   1110.0        1103   1413.0        1418            DL       1641   

        TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  \
0        N712SW              128.0           150.0    116.0     -14.0   
1        N772SW              128.0           145.0    113.0       2.0   
2        N428WN               96.0            90.0     76.0      14.0   
3        N464WN               90.0            90.0     77.0      34.0   
4        N726SW              101.0           115.0     87.0      11.0   
...         ...                ...             ...      ...       ...   
1936753  N938DL              147.0           152.0    120.0      25.0   
1936754  N3743H              127.0           109.0     78.0      75.0   
1936755  N909DA              162.0           143.0    122.0      99.0   
1936756  N646DL              115.0           117.0     89.0       9.0   
1936757  N908DL              123.0           135.0    104.0      -5.0   

         DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  Cancelled  \
0             8.0    IAD  TPA       810     4.0      8.0          0   
1            19.0    IAD  TPA       810     5.0     10.0          0   
2             8.0    IND  BWI       515     3.0     17.0          0   
3            34.0    IND  BWI       515     3.0     10.0          0   
4            25.0    IND  JAX       688     4.0     10.0          0   
...           ...    ...  ...       ...     ...      ...        ...   
1936753      30.0    MSP  ATL       906     9.0     18.0          0   
1936754      57.0    RIC  ATL       481    15.0     34.0          0   
1936755      80.0    ATL  IAH       689     8.0     32.0          0   
1936756      11.0    IAD  ATL       533    13.0     13.0          0   
1936757       7.0    SAT  ATL       874     8.0     11.0          0   

         CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0                 NaN           NaN       NaN            NaN   
1                 NaN           NaN       NaN            NaN   
2                 NaN           NaN       NaN            NaN   
3                 2.0           0.0       0.0            0.0   
4                 NaN           NaN       NaN            NaN   
...               ...           ...       ...            ...   
1936753           3.0           0.0       0.0            0.0   
1936754           0.0          57.0      18.0            0.0   
1936755           1.0           0.0      19.0            0.0   
1936756           NaN           NaN       NaN            NaN   
1936757           NaN           NaN       NaN            NaN   

         LateAircraftDelay  
0                      NaN  
1                      NaN  
2                      NaN  
3                     32.0  
4                      NaN  
...                    ...  
1936753               22.0  
1936754                0.0  
1936755               79.0  
1936756                NaN  
1936757                NaN  

[1936758 rows x 23 columns]

In [10]:
df_flights_new.shape

(1936758, 23)

## Ejercicio 2

### 1. Resume estadísticamente el conjunto de datos y las columnas de interés. Realiza un análisis estadístico de lo que consideres relevante.

Usaremos `.describe()` y `.value_counts` para analizar las columnas:

In [11]:
df_flights_new['DepTime'].describe()

count    1.936758e+06
mean     1.518534e+03
std      4.504853e+02
min      1.000000e+00
25%      1.203000e+03
50%      1.545000e+03
75%      1.900000e+03
max      2.400000e+03
Name: DepTime, dtype: float64

Los números son demasiado grandes, entonces se aplica la función `pd.options.display.float_format` para poder visualizar mejor los datos.

In [12]:
pd.options.display.float_format = '{:.2f}'.format
df_flights_new['DepTime'].describe()

count   1936758.00
mean       1518.53
std         450.49
min           1.00
25%        1203.00
50%        1545.00
75%        1900.00
max        2400.00
Name: DepTime, dtype: float64

Lo más relevante que podemos observar es que la columna _DepTime_ contiene los horarios de la salida de los vuelos registrada como _float_ en vez de horas

In [13]:
df_flights_new[['CRSDepTime','ArrTime', 'CRSArrTime']].describe()

CRSDepTime    ArrTime  CRSArrTime
count  1936758.00 1929648.00  1936758.00
mean      1467.47    1610.14     1634.22
std        424.77     548.18      464.63
min          0.00       1.00        0.00
25%       1135.00    1316.00     1325.00
50%       1510.00    1715.00     1705.00
75%       1815.00    2030.00     2014.00
max       2359.00    2400.00     2400.00

En el dataset había más columnas con horarios, por lo que las juntamos para poder ver toda la información a la vez.  Esto nos dió como resultado que también tienen los valores registrados como _float_.  

Sabiendo esto, podemos cambiar a _horas_ en vez de _float_ los valores de estas columnas, más adelante durante la etapa de limpieza del dataset.

In [14]:
df_flights_new['UniqueCarrier'].value_counts()

WN    377602
AA    191865
MQ    141920
UA    141426
OO    132433
DL    114238
XE    103663
CO    100195
US     98425
EV     81877
NW     79108
FL     71284
YV     67063
B6     55315
OH     52657
9E     51885
AS     39293
F9     28269
HA      7490
AQ       750
Name: UniqueCarrier, dtype: int64

Nada relevante con la columna _UniqueCarrier_

In [15]:
df_flights_new['FlightNum'].value_counts()

16      1586
50      1550
40      1507
15      1471
44      1447
        ... 
7681       1
7032       1
6893       1
6914       1
5272       1
Name: FlightNum, Length: 7499, dtype: int64

In [16]:
df_flights_new['TailNum'].value_counts()

N325SW    965
N676SW    951
N658SW    945
N313SW    937
N308SA    936
         ... 
9189E       1
N853NW      1
N856NW      1
N76010      1
N78008      1
Name: TailNum, Length: 5366, dtype: int64

In [17]:
df_flights_new[['FlightNum', 'TailNum']].value_counts()

FlightNum  TailNum
64         N764AS     122
           N765AS     113
           N763AS     109
           N768AS      99
153        N768AS      98
                     ... 
1211       N311UA       1
           N313US       1
           N314NB       1
           N315NB       1
9742       91669E       1
Length: 1132032, dtype: int64

Si analizamos las columnas de 'FlightNum' y 'TailNum' por separado, lo único relevante que encontramos es que hay tanto vuelos como aviones que se repiten más de 900 veces y otros que solo se usaron 1 vez.

Sin embargo, si los analizamos juntos, podemos observar que diferentes aviones vuelan el mismo número de vuelo _FlightNum_ lo que podría ser relevante más adelante.

In [18]:
df_flights_new[['ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay', 'DepDelay']].describe()

ActualElapsedTime  CRSElapsedTime    AirTime   ArrDelay   DepDelay
count         1928371.00      1936560.00 1928371.00 1928371.00 1936758.00
mean              133.31          134.30     108.28      42.20      43.19
std                72.06           71.34      68.64      56.78      53.40
min                14.00          -25.00       0.00    -109.00       6.00
25%                80.00           82.00      58.00       9.00      12.00
50%               116.00          116.00      90.00      24.00      24.00
75%               165.00          165.00     137.00      56.00      53.00
max              1114.00          660.00    1091.00    2461.00    2467.00

De primera instancia podemos observar una variación estandar muy alta en las 5 columnas; al igual que la diferencia entre el mínimo y el máximo.  Esto seguramente se debe a las distancias que existen entre los aeropuertos de origen y los de destino, pero es importante saberlo porque podría ser relevante más adelante.

In [19]:
df_flights_new['Origin'].value_counts()

ATL    131613
ORD    125979
DFW     95414
DEN     74323
LAX     58772
        ...  
BJI         4
PIR         3
PUB         2
INL         1
TUP         1
Name: Origin, Length: 303, dtype: int64

In [20]:
df_flights_new['Dest'].value_counts()

ORD    108984
ATL    106898
DFW     70657
DEN     63003
LAX     59969
        ...  
INL         9
PIR         3
CYS         1
TUP         1
OGD         1
Name: Dest, Length: 304, dtype: int64

Al analizar el número de aeropuertos de origen y destino, encontramos en que hay una diferencia de 1. Esto lo podremos ver más a detalle en la limpieza de los datos.

In [21]:
df_flights_new[['Origin','Dest']].value_counts()

Origin  Dest
LAX     SFO     4739
ORD     LGA     4396
ATL     LGA     4058
SFO     LAX     4020
LAS     LAX     3516
                ... 
TOL     CVG        1
BZN     IDA        1
ATW     DSM        1
OKC     IAD        1
GEG     LGB        1
Length: 5205, dtype: int64

Y si juntamos el origen y destinos podemos observar que hay rutas con más de 4 mil vuelos y otras con sólo 1.

In [22]:
df_flights_new[['TaxiIn', 'TaxiOut']].describe()

TaxiIn    TaxiOut
count 1929648.00 1936303.00
mean        6.81      18.23
std         5.27      14.34
min         0.00       0.00
25%         4.00      10.00
50%         6.00      14.00
75%         8.00      21.00
max       240.00     422.00

Lo que más llama la atención es que los tiempos máximos son de 240 y 422 minutos respectivamente, pero los mínimos son de 0.  

Se investigó un poco más que significan estas columnas, y lo que descubrimos es que _Taxi In y Out_ es el tiempo en minutos que un avión tarda en desplazarse por tierra desde la puerta de embarque hasta la pista de despegue y viceversa.  Y aunque el tiempo de taxi no se considera parte del tiempo de vuelo real, por lo que generalmente no se incluye en las estadísticas de duración del vuelo, podría ser relevante más adelante.

In [23]:
df_flights_new[df_flights_new['TaxiOut'] >= 150]

DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum  \
29933    2001.00        1800    52.00        2045            WN        103   
34760     758.00         730  1211.00         855            XE       2322   
34955    1504.00        1320  1845.00        1455            XE       2297   
35511    1241.00        1215  1735.00        1437            XE       1284   
35948     941.00         848  1332.00        1018            XE       2639   
...          ...         ...      ...         ...           ...        ...   
1930325    12.00        1914   728.00        2334            CO       1776   
1930337  1828.00        1633  2352.00        1947            CO       1820   
1930340  2007.00        1900   516.00        2228            CO       1830   
1930434  1604.00        1545  1952.00        1704            CO        614   
1934931  1954.00        1910    28.00        2127            DL        846   

        TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  \
29933    N784SW             291.00          165.00   138.00    247.00   
34760    N14542             253.00           85.00    77.00    196.00   
34955    N13964             221.00           95.00    57.00    230.00   
35511    N14998             294.00          142.00   104.00    178.00   
35948    N15986             231.00           90.00    66.00    194.00   
...         ...                ...             ...      ...       ...   
1930325  N14214             376.00          200.00   165.00    474.00   
1930337  N39728             264.00          134.00    89.00    245.00   
1930340  N39418             489.00          148.00   115.00    408.00   
1930434  N14605             228.00           79.00    48.00    168.00   
1934931  N602DL             274.00          137.00   112.00    181.00   

         DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  Cancelled  \
29933      121.00    MDW  HOU       937    3.00   150.00          0   
34760       28.00    MAF  IAH       429    6.00   170.00          0   
34955      104.00    MHT  EWR       209   10.00   154.00          0   
35511       26.00    ATL  EWR       745   12.00   178.00          0   
35948       53.00    OKC  IAH       395   14.00   151.00          0   
...           ...    ...  ...       ...     ...      ...        ...   
1930325    298.00    IAH  PHL      1324    7.00   204.00          0   
1930337    115.00    IAH  ATL       689    7.00   168.00          0   
1930340     67.00    IAH  PBI       956    4.00   370.00          0   
1930434     19.00    DFW  IAH       224   12.00   168.00          0   
1934931     44.00    ATL  EWR       745    6.00   156.00          0   

         CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
29933            0.00          5.00    126.00           0.00   
34760            0.00          0.00    196.00           0.00   
34955            0.00          0.00    230.00           0.00   
35511            0.00          0.00    164.00           0.00   
35948            0.00          0.00    194.00           0.00   
...               ...           ...       ...            ...   
1930325          0.00        298.00    176.00           0.00   
1930337          0.00          0.00    245.00           0.00   
1930340          0.00         67.00    341.00           0.00   
1930434          0.00          0.00    149.00           0.00   
1934931          0.00          0.00    169.00           0.00   

         LateAircraftDelay  
29933               116.00  
34760                 0.00  
34955                 0.00  
35511                14.00  
35948                 0.00  
...                    ...  
1930325               0.00  
1930337               0.00  
1930340               0.00  
1930434              19.00  
1934931              12.00  

[1824 rows x 23 columns]

Al tratar de entender como funcionan _Taxi In_ y _Out_ descubrimos que la suma de _CarrierDelay, WeatherDelay, NASDelay, SecurityDelay y LateAircraftDelay_ son los valores de _ArrDelay_.

In [24]:
df_flights_new[['CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']].describe()

CarrierDelay  WeatherDelay   NASDelay  SecurityDelay  LateAircraftDelay
count    1247488.00    1247488.00 1247488.00     1247488.00         1247488.00
mean          19.18          3.70      15.02           0.09              25.30
std           43.55         21.49      33.83           2.02              42.05
min            0.00          0.00       0.00           0.00               0.00
25%            0.00          0.00       0.00           0.00               0.00
50%            2.00          0.00       2.00           0.00               8.00
75%           21.00          0.00      15.00           0.00              33.00
max         2436.00       1352.00    1357.00         392.00            1316.00

Al analizar las columnas _'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'_ lo que más llama la atención es que los minimos y el 25% percentil está en 0. Por lo que estas son columnas que no son necesarias al tener tantos valores incompletos.  También eliminaremos las columnas de _'TaxiIn'_ y _'TaxiOut'_ pues no encontramos gran relevancia en su información.

Entonces, usamos `.drop()` para deshacernos de esas columnas que son innecesarias.

In [25]:
df_flights_new = df_flights_new.drop(['CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay','TaxiIn','TaxiOut'], axis=1)
df_flights_new

DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum  \
0        2003.00        1955  2211.00        2225            WN        335   
1         754.00         735  1002.00        1000            WN       3231   
2         628.00         620   804.00         750            WN        448   
3        1829.00        1755  1959.00        1925            WN       3920   
4        1940.00        1915  2121.00        2110            WN        378   
...          ...         ...      ...         ...           ...        ...   
1936753  1250.00        1220  1617.00        1552            DL       1621   
1936754   657.00         600   904.00         749            DL       1631   
1936755  1007.00         847  1149.00        1010            DL       1631   
1936756  1251.00        1240  1446.00        1437            DL       1639   
1936757  1110.00        1103  1413.00        1418            DL       1641   

        TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  \
0        N712SW             128.00          150.00   116.00    -14.00   
1        N772SW             128.00          145.00   113.00      2.00   
2        N428WN              96.00           90.00    76.00     14.00   
3        N464WN              90.00           90.00    77.00     34.00   
4        N726SW             101.00          115.00    87.00     11.00   
...         ...                ...             ...      ...       ...   
1936753  N938DL             147.00          152.00   120.00     25.00   
1936754  N3743H             127.00          109.00    78.00     75.00   
1936755  N909DA             162.00          143.00   122.00     99.00   
1936756  N646DL             115.00          117.00    89.00      9.00   
1936757  N908DL             123.00          135.00   104.00     -5.00   

         DepDelay Origin Dest  Distance  Cancelled  
0            8.00    IAD  TPA       810          0  
1           19.00    IAD  TPA       810          0  
2            8.00    IND  BWI       515          0  
3           34.00    IND  BWI       515          0  
4           25.00    IND  JAX       688          0  
...           ...    ...  ...       ...        ...  
1936753     30.00    MSP  ATL       906          0  
1936754     57.00    RIC  ATL       481          0  
1936755     80.00    ATL  IAH       689          0  
1936756     11.00    IAD  ATL       533          0  
1936757      7.00    SAT  ATL       874          0  

[1936758 rows x 16 columns]

### 2. Encuentra cuántos datos faltantes hay por columna.

Para saber cuantos datos faltan por columna, usamos `isna()` que nos dice cuantos valores ausentes hay en cada columna, pero en forma de tabla y dificil de leer al usar un dataset tan grande.  Para solucionarlo, agregamos el método `.sum()`.

In [26]:
df_flights_new.isna().sum()

DepTime                 0
CRSDepTime              0
ArrTime              7110
CRSArrTime              0
UniqueCarrier           0
FlightNum               0
TailNum                 5
ActualElapsedTime    8387
CRSElapsedTime        198
AirTime              8387
ArrDelay             8387
DepDelay                0
Origin                  0
Dest                    0
Distance                0
Cancelled               0
dtype: int64

Una vez que sabemos que datos faltan en cada columna, comenzaremos a limpiarlos uno por uno comenzando con los más pequeños.  

Primero crearemos una copia del data frame `df_flights_new` para poder hacer los cambios tranquilamente.

In [27]:
df_flights_copia = df_flights_new.copy()
df_flights_copia

DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum  \
0        2003.00        1955  2211.00        2225            WN        335   
1         754.00         735  1002.00        1000            WN       3231   
2         628.00         620   804.00         750            WN        448   
3        1829.00        1755  1959.00        1925            WN       3920   
4        1940.00        1915  2121.00        2110            WN        378   
...          ...         ...      ...         ...           ...        ...   
1936753  1250.00        1220  1617.00        1552            DL       1621   
1936754   657.00         600   904.00         749            DL       1631   
1936755  1007.00         847  1149.00        1010            DL       1631   
1936756  1251.00        1240  1446.00        1437            DL       1639   
1936757  1110.00        1103  1413.00        1418            DL       1641   

        TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  \
0        N712SW             128.00          150.00   116.00    -14.00   
1        N772SW             128.00          145.00   113.00      2.00   
2        N428WN              96.00           90.00    76.00     14.00   
3        N464WN              90.00           90.00    77.00     34.00   
4        N726SW             101.00          115.00    87.00     11.00   
...         ...                ...             ...      ...       ...   
1936753  N938DL             147.00          152.00   120.00     25.00   
1936754  N3743H             127.00          109.00    78.00     75.00   
1936755  N909DA             162.00          143.00   122.00     99.00   
1936756  N646DL             115.00          117.00    89.00      9.00   
1936757  N908DL             123.00          135.00   104.00     -5.00   

         DepDelay Origin Dest  Distance  Cancelled  
0            8.00    IAD  TPA       810          0  
1           19.00    IAD  TPA       810          0  
2            8.00    IND  BWI       515          0  
3           34.00    IND  BWI       515          0  
4           25.00    IND  JAX       688          0  
...           ...    ...  ...       ...        ...  
1936753     30.00    MSP  ATL       906          0  
1936754     57.00    RIC  ATL       481          0  
1936755     80.00    ATL  IAH       689          0  
1936756     11.00    IAD  ATL       533          0  
1936757      7.00    SAT  ATL       874          0  

[1936758 rows x 16 columns]

En la columna _TailNum_ rellenaremos los datos _NaN_ con **"No disponible"** usando `.fillna()` y después comprobaremos que todo haya salido bien.

In [28]:
df_flights_copia['TailNum'].fillna('No disponible', inplace=True)
df_flights_copia[df_flights_copia['TailNum'] == 'No disponible']

DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum  \
433449   1333.00        1320      NaN        1428            OH       5338   
523748   1716.00        1655  2036.00        1947            9E       3760   
773772   1545.00        1530  1711.00        1707            OH       5043   
776480   1720.00        1244  1920.00        1427            OH       5396   
1309410  1840.00        1305      NaN        1438            OH       5731   

               TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  \
433449   No disponible                NaN           68.00      NaN       NaN   
523748   No disponible             140.00          112.00   100.00     49.00   
773772   No disponible              86.00           97.00    66.00      4.00   
776480   No disponible             120.00          103.00    72.00    293.00   
1309410  No disponible                NaN           93.00      NaN       NaN   

         DepDelay Origin Dest  Distance  Cancelled  
433449      13.00    CVG  GRR       268          0  
523748      21.00    DSM  DTW       534          0  
773772      15.00    CVG  CHS       497          0  
776480     276.00    RDU  JFK       426          0  
1309410    335.00    RDU  ATL       356          0

In [29]:
df_flights_copia['TailNum'].isna().sum()

0

La siguiente columna con menos valores vacíos (198), es _CRSElapsedTime_.  Nos dimos cuenta que esta columna también se relaciona con _ActualElapsedTime_ y con _AirTime_, pero por si solas no nos dicen nada pues coinciden en los espacios vacios. Es decir, las valores vacíos de _CRSElapsedTime_ son los mismos que en _ActualElapsedTime_, _ArrDelay_ y con _AirTime_.

Al  hacer un análisis de todas las columnas, se llegó a la conclusión que lo que tenían en común era la columna de _Distance_ y a partir de ahí se pudo sacar la media de estas columnas usando `.groupby()` y `mean()`.

In [30]:
means = df_flights_copia.groupby('Distance')[['CRSElapsedTime', 'ActualElapsedTime', 'ArrDelay', 'AirTime']].mean()
means

CRSElapsedTime  ActualElapsedTime  ArrDelay  AirTime
Distance                                                      
11                 60.00              54.00    100.00    22.50
24                 38.50              40.50     84.00    10.50
30                 52.50              73.25     35.75    20.00
31                 26.17              24.82     44.90    12.51
35                 34.00              27.50     14.00    12.50
...                  ...                ...       ...      ...
4184              537.73             530.95     28.31   502.04
4213              472.44             470.50     42.08   451.48
4243              528.83             524.40     38.35   498.51
4502              565.24             560.83     36.02   532.38
4962              632.38             626.84     32.22   596.25

[1419 rows x 4 columns]

Teniendo esta información, se pueden rellenar las columnas: 
* _AirTime_
* _ActualElapsedTime_
* _ArrDelay_
* _CRSElapsedTime_

Para lograrlo se usaron `.fillna()` y `map`.

In [31]:
df_flights_copia['AirTime'].fillna(df_flights_copia['Distance'].map(means['AirTime']), inplace=True)
df_flights_copia['ActualElapsedTime'].fillna(df_flights_copia['Distance'].map(means['ActualElapsedTime']), inplace=True)
df_flights_copia['ArrDelay'].fillna(df_flights_copia['Distance'].map(means['ArrDelay']), inplace=True)
df_flights_copia['CRSElapsedTime'].fillna(df_flights_copia['Distance'].map(means['CRSElapsedTime']), inplace=True)
df_flights_copia.isna().sum()

DepTime                 0
CRSDepTime              0
ArrTime              7110
CRSArrTime              0
UniqueCarrier           0
FlightNum               0
TailNum                 0
ActualElapsedTime      15
CRSElapsedTime          2
AirTime                15
ArrDelay               15
DepDelay                0
Origin                  0
Dest                    0
Distance                0
Cancelled               0
dtype: int64

Para completar las columnas _ArrTime_, se toma el percentil 75 pues así nos aseguramos de usar los valores más usados.  

Usaremos `.quantile(0.75)` para poder hacerlo.

In [32]:
df_flights_copia['ArrTime'].fillna(df_flights_copia['ArrTime'].quantile(0.75), inplace=True)
df_flights_copia.isna().sum()

DepTime               0
CRSDepTime            0
ArrTime               0
CRSArrTime            0
UniqueCarrier         0
FlightNum             0
TailNum               0
ActualElapsedTime    15
CRSElapsedTime        2
AirTime              15
ArrDelay             15
DepDelay              0
Origin                0
Dest                  0
Distance              0
Cancelled             0
dtype: int64

Finalmente, los datos que no se pudieron completar con media, mediana, percentil, etc., se completaran con `.interpolate` pues son muy pequeños y no afectan en nada a la base de datos.

In [33]:
df_flights_copia['ActualElapsedTime'] = df_flights_copia['ActualElapsedTime'].interpolate()
df_flights_copia['CRSElapsedTime'] = df_flights_copia['CRSElapsedTime'].interpolate()
df_flights_copia['AirTime'] = df_flights_copia['AirTime'].interpolate()
df_flights_copia['ArrDelay'] = df_flights_copia['ArrDelay'].interpolate()
df_flights_copia.isna().sum()

DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
Cancelled            0
dtype: int64

### Reduce la dimensión del dataset de manera aleatoria para obtener un dataset de solo 200,000 registros. Todos los ejercicios se deben hacer con este dataset reducido.

Una vez limpio el dataset, se usa `.sample()` para reducirlo a 200,00 registros.

In [34]:
df_flights_copia = df_flights_copia.sample(n=200000, random_state=1)
df_flights_copia

DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum  \
1750755  1332.00        1320  1436.00        1430            WN       3858   
1074983  1731.00        1720  2013.00        1955            CO        268   
46126     735.00         645   918.00         815            YV       7343   
327441   1333.00        1230  1515.00        1438            9E       2927   
1363177  1359.00        1346  1517.00        1447            FL        301   
...          ...         ...      ...         ...           ...        ...   
1595898  2159.00        1920   112.00        2225            MQ       4205   
1503090  2052.00        1930   115.00        2245            AA        346   
742602   1335.00        1325  1422.00        1415            WN       1902   
1840527   720.00         710   839.00         831            DL       1653   
1526456  1657.00        1650  2217.00        2205            WN        748   

        TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  \
1750755  N700GS              64.00           70.00    49.00      6.00   
1074983  N54241             342.00          335.00   316.00     18.00   
46126    N27185             103.00           90.00    54.00     63.00   
327441   87189E             102.00          128.00    80.00     37.00   
1363177  N267AT             198.00          181.00   178.00     30.00   
...         ...                ...             ...      ...       ...   
1595898  N643MQ             133.00          125.00   109.00    167.00   
1503090  N433AA             203.00          135.00   115.00    150.00   
742602   N316SW              47.00           50.00    37.00      7.00   
1840527  N996DL              79.00           81.00    50.00      8.00   
1526456  N675AA             200.00          195.00   185.00     12.00   

         DepDelay Origin Dest  Distance  Cancelled  
1750755     12.00    CLE  BWI       314          0  
1074983     11.00    EWR  LAS      2227          0  
46126       50.00    ORD  BNA       409          0  
327441      63.00    ORF  DTW       529          0  
1363177     13.00    ATL  DEN      1199          0  
...           ...    ...  ...       ...        ...  
1595898    159.00    ORD  BDL       783          0  
1503090     82.00    ORD  LGA       733          0  
742602      10.00    OKC  DAL       181          0  
1840527     10.00    CLT  ATL       227          0  
1526456      7.00    PHX  MDW      1444          0  

[200000 rows x 16 columns]

### 3. Crea nuevas columnas de Velocidad Media del vuelo y si ha llegado tarde o no.

#### Columna de "Velocidad Media".

Para hacerlo se crea una la columna _'Velocidad_Media'_ dividiendo los valores de _'AirTime_ dividido entre 60 - para convertir los minutos a horas - y entre la Distancia para sacar la velocidad promedio.

In [35]:
df_flights_copia['Velocidad_Media'] = df_flights_copia['Distance'] / (df_flights_copia['AirTime'] / 60)
df_flights_copia

DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum  \
1750755  1332.00        1320  1436.00        1430            WN       3858   
1074983  1731.00        1720  2013.00        1955            CO        268   
46126     735.00         645   918.00         815            YV       7343   
327441   1333.00        1230  1515.00        1438            9E       2927   
1363177  1359.00        1346  1517.00        1447            FL        301   
...          ...         ...      ...         ...           ...        ...   
1595898  2159.00        1920   112.00        2225            MQ       4205   
1503090  2052.00        1930   115.00        2245            AA        346   
742602   1335.00        1325  1422.00        1415            WN       1902   
1840527   720.00         710   839.00         831            DL       1653   
1526456  1657.00        1650  2217.00        2205            WN        748   

        TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  \
1750755  N700GS              64.00           70.00    49.00      6.00   
1074983  N54241             342.00          335.00   316.00     18.00   
46126    N27185             103.00           90.00    54.00     63.00   
327441   87189E             102.00          128.00    80.00     37.00   
1363177  N267AT             198.00          181.00   178.00     30.00   
...         ...                ...             ...      ...       ...   
1595898  N643MQ             133.00          125.00   109.00    167.00   
1503090  N433AA             203.00          135.00   115.00    150.00   
742602   N316SW              47.00           50.00    37.00      7.00   
1840527  N996DL              79.00           81.00    50.00      8.00   
1526456  N675AA             200.00          195.00   185.00     12.00   

         DepDelay Origin Dest  Distance  Cancelled  Velocidad_Media  
1750755     12.00    CLE  BWI       314          0           384.49  
1074983     11.00    EWR  LAS      2227          0           422.85  
46126       50.00    ORD  BNA       409          0           454.44  
327441      63.00    ORF  DTW       529          0           396.75  
1363177     13.00    ATL  DEN      1199          0           404.16  
...           ...    ...  ...       ...        ...              ...  
1595898    159.00    ORD  BDL       783          0           431.01  
1503090     82.00    ORD  LGA       733          0           382.43  
742602      10.00    OKC  DAL       181          0           293.51  
1840527     10.00    CLT  ATL       227          0           272.40  
1526456      7.00    PHX  MDW      1444          0           468.32  

[200000 rows x 17 columns]

#### Columna de "Arribo"
Para hacerlo se usa la columna _Cancelled_ para saber si un vuelo arribó a su destino o no.  Esto se hace usando el método de `.apply()` y la función `lambda`.

In [36]:
df_flights_copia['Arribo'] = df_flights_copia['Cancelled'].apply(lambda x: 'No' if x == 1 else 'Sí')
df_flights_copia

DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum  \
1750755  1332.00        1320  1436.00        1430            WN       3858   
1074983  1731.00        1720  2013.00        1955            CO        268   
46126     735.00         645   918.00         815            YV       7343   
327441   1333.00        1230  1515.00        1438            9E       2927   
1363177  1359.00        1346  1517.00        1447            FL        301   
...          ...         ...      ...         ...           ...        ...   
1595898  2159.00        1920   112.00        2225            MQ       4205   
1503090  2052.00        1930   115.00        2245            AA        346   
742602   1335.00        1325  1422.00        1415            WN       1902   
1840527   720.00         710   839.00         831            DL       1653   
1526456  1657.00        1650  2217.00        2205            WN        748   

        TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  \
1750755  N700GS              64.00           70.00    49.00      6.00   
1074983  N54241             342.00          335.00   316.00     18.00   
46126    N27185             103.00           90.00    54.00     63.00   
327441   87189E             102.00          128.00    80.00     37.00   
1363177  N267AT             198.00          181.00   178.00     30.00   
...         ...                ...             ...      ...       ...   
1595898  N643MQ             133.00          125.00   109.00    167.00   
1503090  N433AA             203.00          135.00   115.00    150.00   
742602   N316SW              47.00           50.00    37.00      7.00   
1840527  N996DL              79.00           81.00    50.00      8.00   
1526456  N675AA             200.00          195.00   185.00     12.00   

         DepDelay Origin Dest  Distance  Cancelled  Velocidad_Media Arribo  
1750755     12.00    CLE  BWI       314          0           384.49     Sí  
1074983     11.00    EWR  LAS      2227          0           422.85     Sí  
46126       50.00    ORD  BNA       409          0           454.44     Sí  
327441      63.00    ORF  DTW       529          0           396.75     Sí  
1363177     13.00    ATL  DEN      1199          0           404.16     Sí  
...           ...    ...  ...       ...        ...              ...    ...  
1595898    159.00    ORD  BDL       783          0           431.01     Sí  
1503090     82.00    ORD  LGA       733          0           382.43     Sí  
742602      10.00    OKC  DAL       181          0           293.51     Sí  
1840527     10.00    CLT  ATL       227          0           272.40     Sí  
1526456      7.00    PHX  MDW      1444          0           468.32     Sí  

[200000 rows x 18 columns]

### 4. Haz una tabla de las aerolíneas con más retrasos acumulados.

Para hacerlo usamos `.groupby()`,`.sum()` y `.sort_values()` para crear la agrupación de las aerolineas sumando los valores de _'ArrDelay'_ para así saber el número de minutos acumulados en los retrasos. Finalmente se ordenan de mayor a menor y se usa `pd.DataFrame()` para crear el dataframe.

In [49]:
pd.DataFrame(df_flights_copia.groupby('UniqueCarrier')[['ArrDelay','DepDelay']].sum().sort_values(by='ArrDelay', ascending=False).reset_index())

UniqueCarrier   ArrDelay   DepDelay
0             WN 1175079.35 1355311.00
1             AA  918237.12  920503.00
2             UA  687496.54  718663.00
3             MQ  659183.70  641468.00
4             OO  624390.41  617910.00
5             XE  534838.92  531367.00
6             DL  475659.42  465592.00
7             CO  431183.12  461822.00
8             EV  395122.08  402386.00
9             YV  386287.47  386332.00
10            US  379951.19  402989.00
11            NW  356927.52  337015.00
12            FL  325325.78  316232.00
13            B6  308821.92  309763.00
14            OH  271080.46  259925.00
15            9E  248647.13  251327.00
16            AS  145497.98  154595.00
17            F9   88667.78   87874.00
18            HA   26567.41   25493.00
19            AQ    1580.00    1753.00

### 5.1 ¿Cuáles son los vuelos más largos?

Para hacerlo usamos el método `.sort_values()`con la especificación `ascending=False` y también `.drop_duplicates()` para evitar que  el dataframe fuera tan largo.

In [99]:
vuelos_largos = df_flights_copia[['FlightNum', 'Distance', 'Origin','Dest']].drop_duplicates().sort_values(by='Distance', ascending=False).reset_index(drop=True)
vuelos_largos

FlightNum  Distance Origin Dest
0             14      4962    HNL  EWR
1             15      4962    EWR  HNL
2           1560      4502    HNL  ATL
3            850      4502    HNL  ATL
4           1273      4502    ATL  HNL
...          ...       ...    ...  ...
56302       5471        36    SNA  LAX
56303       5471        36    LAX  SNA
56304         65        31    WRG  PSG
56305         64        31    PSG  WRG
56306       2009        24    IAH  HOU

[56307 rows x 4 columns]

### 5.2 ¿Y los que tienen más retrasados?

Para esta filtración, también usamos `groupby()` y `sort_values()` como el código anterior, pero ademas el método `sum()` y `.agg()` para agregar las columnas de _Origin_ y _Dest_ y así tener una mejor presentación.

In [100]:
vuelos_delay = df_flights_copia.groupby('FlightNum').agg({'ArrDelay':'sum', 'DepDelay':'sum', 'Origin':'first', 'Dest':'last'}).sort_values(['ArrDelay','DepDelay','Origin','Dest'], ascending=False).reset_index(drop=True)
vuelos_delay

ArrDelay  DepDelay Origin Dest
0      8365.74   8803.00    HOU  OGG
1      8126.13   8158.00    HNL  SFO
2      7715.00   7949.00    SFO  CVG
3      7605.81   7367.00    DAL  ELP
4      7597.21   7636.00    EWR  ORD
...        ...       ...    ...  ...
7279    -11.00      7.00    SYR  IAD
7280    -11.00      7.00    DAY  ORD
7281    -15.00     33.00    JFK  LNK
7282    -20.00     26.00    EUG  DEN
7283    -20.00     20.00    JFK  BGR

[7284 rows x 4 columns]

### 5.3 Busca las rutas más largas

Aquí también se usaron `.sort_values()` y `.drop_duplicates()`

In [101]:
rutas_largas = df_flights_copia[['Origin', 'Dest', 'Distance']].drop_duplicates().sort_values(by='Distance', ascending=False).reset_index(drop=True)
rutas_largas

Origin Dest  Distance
0       HNL  EWR      4962
1       EWR  HNL      4962
2       HNL  ATL      4502
3       ATL  HNL      4502
4       ORD  HNL      4243
...     ...  ...       ...
4800    SNA  LAX        36
4801    LAX  SNA        36
4802    PSG  WRG        31
4803    WRG  PSG        31
4804    IAH  HOU        24

[4805 rows x 3 columns]

### 5.4 Y las que acumulen más retardos

In [81]:
df_flights_copia.groupby(['Origin', 'Dest'])[['ArrDelay', 'DepDelay']].agg('sum').drop_duplicates().sort_values(by='ArrDelay', ascending=False)


ArrDelay  DepDelay
Origin Dest                    
ORD    LGA   26038.63  24790.00
LAX    SFO   23462.00  23977.00
LGA    ORD   23278.72  22879.00
ORD    EWR   22132.00  24540.00
ATL    EWR   22040.89  21772.00
...               ...       ...
CAK    LAS     -25.00     41.00
SBA    SBP     -33.00     11.00
GEG    TUS     -34.00     47.00
SMF    OGG     -40.00      6.00
HNL    ANC    -170.00     62.00

[4780 rows x 2 columns]

### Proporciona lo que consideres relevante.

¿Cual es la correlación entre _Tiempo de Vuelo, Distancia y Velocidad Media_?

In [102]:
df_flights_copia[['AirTime', 'Distance', 'Velocidad_Media']].corr()

AirTime  Distance  Velocidad_Media
AirTime             1.00      0.98             0.53
Distance            0.98      1.00             0.63
Velocidad_Media     0.53      0.63             1.00

## Ejercicio 3

### Exporta el dataset limpio y con las nuevas columnas a Excel.

In [103]:
df_flights_copia.to_excel('vuelos.xlsx')

# FIN